Switch to scala 2.11

davidegrohmann opened the pull request:
Refactor the code generation part of the new runtime in order to avoid inheritance across Java and Scala which causes compilation errors.

davidegrohmann committed:
Switch to scala 2.11

Refactor the code generation part of the new runtime in order to avoid
inheritance across Java and Scala which causes compilation errors.

davidegrohmann commented on community/cypher/cypher/pom.xml:
```diff
@@ -169,8 +169,9 @@
-      <version>2.0.4</version>
+      <artifactId>neo4j-cypher-compiler-1.9_2.11</artifactId>
+      <!-- FIXME: this should be replaced by the 2.0.5 released 1.9 compiler -->
+      <version>2.0.5-SNAPSHOT</version>
```
Replace with a stable release when ready.

davidegrohmann commented:
@thobe  I had to restructure a bit the `CompiledExecutionResult` used by the new runtime and the code generation as well since having a Java class extending a Scala class was causing compilation errors. The current solution is probably sub-optimal, but perhaps good enough to allow us for merging this PR in and cleanup later. WDYT?

thobe commented on community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/executionplan/ToCompileCode.java:
```diff
@@ -0,0 +1,39 @@
+import org.neo4j.cypher.internal.compiler.v2_3.planDescription.InternalPlanDescription;
+import org.neo4j.graphdb.Result;
+
+public interface ToCompileCode
```
The name of this interface is awful!

davidegrohmann replied:
I agree, not proud of that naming. If you have any good suggestion please let me know.

thobe replied:
`GeneratedQueryExecution` or something like that...

thobe commented:
@davidegrohmann The change to have a simple interface implemented by the generated code and then a wrapper to add the support that we want for the tests that are written in scala is sensible, it is what we've been thinking we should do eventually anyhow. We would want to move towards the interface implemented by the generated code being the interface expected by the APIs in kernel, so that there is no indirection in the common case (in the tests in scala we can afford the indirection, but when actually executing queries we would prefer to be without it), but this is something we will refactor towards later, so it's acceptable that this is not done yet as part of this commit, we can fix that later. The name of the interface you introduced is absolutely horrible though, which makes me think that perhaps we should just use the correct interface now, to avoid having to name this short-lived interface.

boggle committed:
Provide old commons classes for use by cypher-compiler-1.9

Ensure binary compatibility with CollectionSupport in
cypher-compiler-2.2 since the two clash

boggle committed:
Adds compatibility handling for 1.9 plan description

boggle committed:
Add parser combinators dependency for 1.9 compiler

pontusmelke commented:
I think it would like to change so that the generated java class only extends pure java stuff, but that is for a later day. However you have lots of licenses fun left…

boggle committed:
Make sure Cypher 1.9 is binary compatible with Neo4j 2.3

unoexperto commented:
Hey guys,

Any ETA on implementing this ?

Thanks.

jotomo commented:
@cppexpert The latest milestone (2.3.0-M02) is using Scala 2.11 already :-)

unoexperto commented:
My god ! I'm so going to try it now. And when is release scheduled ?
